staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -660,6 +695,134 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
 	return nil
 }
 
+// Saves the last-applied-configuration annotation in a separate SSA field manager
+// to prevent it from being dropped by users who have transitioned to SSA.
+//
+// If this operation is not performed, then the last-applied-configuration annotation
+// would be removed from the object upon the first SSA usage.
+func (o *ApplyOptions) saveLastApplyAnnotationIfNecessary(
+	helper *resource.Helper,
+	obj runtime.Object,
+) (runtime.Object, error) {
+	oldFieldManager := helper.FieldManager
+	defer helper.WithFieldManager(oldFieldManager)
+
+	helper = helper.WithFieldManager(FieldManagerClientSideApply)
+
+	// Send an apply patch with the last-applied-annotation
+	// so that it is not orphaned by SSA in the following patch:
+	// First serialize the object without the annotation to prevent recursion,
+	// then add that serialization to it as the annotation and serialize it again.
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return obj, err
+	}
+
+	// Before performing migration, ensure that the last-applied-configuration
+	// annotation is owned by a separate field manager so that it does not get
+	// orphaned and pruned by the server.
+	lastAppliedSeparated := false
+
+	for _, entry := range accessor.GetManagedFields() {
+		if entry.Operation == "Apply" && entry.Manager == FieldManagerClientSideApply {
+			// Operation already undertaken.
+			lastAppliedSeparated = true
+			break
+		}
+	}
+
+	// This operastion is already completed. Nothing to do
+	if lastAppliedSeparated {
+		return obj, nil
+	}
+
+	// Otherwise, use the server side version of the object.
+	// Get the current annotations from the object.
+	annots := accessor.GetAnnotations()
+	if annots == nil {
+		annots = map[string]string{}
+	}
+
+	original := annots[v1.LastAppliedConfigAnnotation]
+
+	justAnnotation := &unstructured.Unstructured{}
+	justAnnotation.SetGroupVersionKind(obj.GetObjectKind().GroupVersionKind())
+	justAnnotation.SetName(accessor.GetName())
+	justAnnotation.SetNamespace(accessor.GetNamespace())
+	justAnnotation.SetAnnotations(map[string]string{
+		v1.LastAppliedConfigAnnotation: original,
+	})
+
+	modified, err := runtime.Encode(unstructured.UnstructuredJSONScheme, justAnnotation)
+	if err != nil {
+		return obj, nil
+	}
+
+	newObj, err := helper.Patch(
+		accessor.GetNamespace(),
+		accessor.GetName(),
+		types.ApplyPatchType,
+		modified,
+		&metav1.PatchOptions{
+			FieldManager: "kubectl-last-applied",
```

Comment:
Maybe use a constant for this (I like its name), and also that's not the name you're looking for at the beginning of the function?

Reply:
yeah this field manager value is ignored. I was trying to explain it gets overridden by the helper